File: src-tauri/src/download.rs
use anyhow::{bail, Context, Result};
use futures_util::StreamExt;
use reqwest::header::{CONTENT_LENGTH, RANGE};
use reqwest::Client;
use serde::Serialize;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use tokio::fs::{self, File, OpenOptions};
use tokio::io::AsyncWriteExt;

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Progress {
    pub id: String,
    pub done: u64,
    pub total: u64,
    pub bytes_per_sec: u64,
}

pub struct Cancel(pub Arc<AtomicBool>);

impl Cancel {
    pub fn stopped(&self) -> bool {
        self.0.load(Ordering::Relaxed)
    }
}

/// Качает файл целиком (с докачкой), сообщая о прогрессе через `on_progress`.
pub async fn fetch<F>(
    http: &Client,
    url: &str,
    dest: &Path,
    cancel: &Cancel,
    mut on_progress: F,
) -> Result<PathBuf>
where
    F: FnMut(u64, u64, u64),
{
    if let Some(dir) = dest.parent() {
        fs::create_dir_all(dir).await?;
    }

    let total = remote_size(http, url).await;

    if let Ok(m) = fs::metadata(dest).await {
        if total.is_none() || Some(m.len()) == total {
            return Ok(dest.to_path_buf());
        }
    }

    let part = dest.with_extension("mp4.part");
    let mut done = fs::metadata(&part).await.map(|m| m.len()).unwrap_or(0);
    // Битый огрызок больше исходника перекачиваем с нуля.
    if let (Some(t), true) = (total, done > 0) {
        if done > t {
            fs::remove_file(&part).await?;
            done = 0;
        }
    }

    if Some(done) == total && done > 0 {
        replace(&part, dest).await?;
        return Ok(dest.to_path_buf());
    }

    let mut req = http
        .get(url)
        .header("Accept", "video/webm,video/ogg,video/*;q=0.9,*/*;q=0.5")
        .header("Referer", crate::site::ORIGIN)
        .header("Sec-Fetch-Dest", "video")
        .header("Sec-Fetch-Mode", "no-cors")
        .header("Sec-Fetch-Site", "cross-site");
    if done > 0 {
        req = req.header(RANGE, format!("bytes={done}-"));
    }

    let resp = req.send().await.context("не удалось начать загрузку")?;
    if !resp.status().is_success() {
        bail!("сервер видео ответил {}", resp.status());
    }
    // Если докачка не поддержана — начинаем заново.
    let resume = done > 0 && resp.status().as_u16() == 206;
    let start = if resume { done } else { 0 };
    let total = total.unwrap_or_else(|| {
        start
            + resp
                .headers()
                .get(CONTENT_LENGTH)
                .and_then(num)
                .unwrap_or(0)
    });

    let mut file = if resume {
        OpenOptions::new().append(true).open(&part).await?
    } else {
        File::create(&part).await?
    };

    let mut written = start;
    let mut stream = resp.bytes_stream();
    let mut tick = std::time::Instant::now();
    let mut tick_bytes = written;
    on_progress(written, total, 0);

    while let Some(chunk) = stream.next().await {
        if cancel.stopped() {
            file.flush().await?;
            bail!("загрузка отменена");
        }
        let chunk = chunk.context("обрыв загрузки")?;
        file.write_all(&chunk).await?;
        written += chunk.len() as u64;

        let elapsed = tick.elapsed();
        if elapsed.as_millis() >= 250 {
            let speed = (written - tick_bytes) as f64 / elapsed.as_secs_f64();
            on_progress(written, total, speed as u64);
            tick = std::time::Instant::now();
            tick_bytes = written;
        }
    }
    file.flush().await?;
    drop(file);

    if total > 0 && written < total {
        bail!("скачано {written} из {total} байт — запустите ещё раз, докачается");
    }
    on_progress(written, total.max(written), 0);
    replace(&part, dest).await?;
    Ok(dest.to_path_buf())
}

/// На Windows rename не перезаписывает существующий файл, в отличие от unix.
async fn replace(from: &Path, to: &Path) -> Result<()> {
    if cfg!(windows) && to.exists() {
        fs::remove_file(to).await?;
    }
    fs::rename(from, to).await?;
    Ok(())
}

fn num(v: &reqwest::header::HeaderValue) -> Option<u64> {
    v.to_str().ok()?.parse().ok()
}

async fn remote_size(http: &Client, url: &str) -> Option<u64> {
    let r = http
        .head(url)
        .header("Referer", crate::site::ORIGIN)
        .send()
        .await
        .ok()?;
    if !r.status().is_success() {
        return None;
    }
    r.headers().get(CONTENT_LENGTH).and_then(num)
}

File: server/download.ts
import fs from "node:fs";
import fsp from "node:fs/promises";
import path from "node:path";
import { msg } from "./schema.ts";
import { UA } from "./site.ts";

export type OnProgress = (done: number, total: number, bytesPerSec: number) => void;

/// На Windows rename не перезаписывает существующий файл, в отличие от unix.
async function replace(from: string, to: string): Promise<void> {
  if (process.platform === "win32") await fsp.rm(to, { force: true });
  await fsp.rename(from, to);
}

const size = async (file: string): Promise<number | null> => {
  try {
    return (await fsp.stat(file)).size;
  } catch {
    return null;
  }
};

/// Пустой Referer шлём именно как отсутствие заголовка: CDN AllVideo режет чужие.
const withReferer = (headers: Record<string, string>, referer: string): Record<string, string> =>
  referer === "" ? headers : { ...headers, Referer: referer };

/// HEAD на sibnet и filevideo отдаёт 403, поэтому размер спрашиваем диапазоном.
async function remoteSize(url: string, referer: string): Promise<number | null> {
  try {
    const res = await fetch(url, {
      headers: withReferer({ "User-Agent": UA, Range: "bytes=0-0" }, referer),
      signal: AbortSignal.timeout(30_000),
    });
    await res.body?.cancel();
    if (!res.ok) return null;
    // `bytes 0-0/12345` при 206; если диапазон проигнорировали — это весь файл.
    const range = res.headers.get("content-range");
    if (range !== null) {
      const n = Number(range.split("/").pop());
      if (Number.isFinite(n) && n > 0) return n;
    }
    if (res.status === 200) {
      const n = Number(res.headers.get("content-length"));
      if (Number.isFinite(n) && n > 0) return n;
    }
    return null;
  } catch {
    return null;
  }
}

/// Качает файл целиком (с докачкой), сообщая о прогрессе через `onProgress`.
export async function fetchFile(
  url: string,
  referer: string,
  dest: string,
  signal: AbortSignal,
  onProgress: OnProgress,
): Promise<string> {
  await fsp.mkdir(path.dirname(dest), { recursive: true });

  const known = await remoteSize(url, referer);

  const already = await size(dest);
  if (already !== null && (known === null || already === known)) return dest;

  const part = `${dest}.part`;
  let done = (await size(part)) ?? 0;
  // Битый огрызок больше исходника перекачиваем с нуля.
  if (known !== null && done > known) {
    await fsp.rm(part, { force: true });
    done = 0;
  }
  if (known !== null && done === known && done > 0) {
    await replace(part, dest);
    return dest;
  }

  const headers: Record<string, string> = withReferer(
    {
      "User-Agent": UA,
      Accept: "video/webm,video/ogg,video/*;q=0.9,*/*;q=0.5",
      "Sec-Fetch-Dest": "video",
      "Sec-Fetch-Mode": "no-cors",
      "Sec-Fetch-Site": "cross-site",
    },
    referer,
  );
  if (done > 0) headers.Range = `bytes=${done}-`;

  let res: Response;
  try {
    res = await fetch(url, { headers, signal });
  } catch (e) {
    if (signal.aborted) throw new Error("загрузка отменена");
    throw new Error(`не удалось начать загрузку: ${msg(e)}`);
  }
  if (!res.ok) {
    await res.body?.cancel();
    throw new Error(`сервер видео ответил ${res.status}`);
  }

  // Если докачка не поддержана — начинаем заново.
  const resume = done > 0 && res.status === 206;
  const start = resume ? done : 0;
  const total = known ?? start + Number(res.headers.get("content-length") ?? 0);

  // FileSink Bun'а вдвое дешевле по CPU, чем node-стрим, — на слабом сервере это заметно.
  const fd = fs.openSync(part, resume ? "a" : "w");
  const out = Bun.file(fd).writer({ highWaterMark: 1 << 20 });
  let written = start;
  let tick = Date.now();
  let tickBytes = written;
  onProgress(written, total, 0);

  try {
    for await (const chunk of res.body as unknown as AsyncIterable<Uint8Array>) {
      out.write(chunk);
      written += chunk.length;
      const elapsed = Date.now() - tick;
      if (elapsed >= 250) {
        onProgress(written, total, Math.round(((written - tickBytes) * 1000) / elapsed));
        tick = Date.now();
        tickBytes = written;
      }
    }
  } catch (e) {
    if (signal.aborted) throw new Error("загрузка отменена");
    throw new Error(`обрыв загрузки: ${msg(e)}`);
  } finally {
    await out.end();
    fs.closeSync(fd);
  }

  if (total > 0 && written < total) {
    throw new Error(`скачано ${written} из ${total} байт — запустите ещё раз, докачается`);
  }
  onProgress(written, Math.max(total, written), 0);
  await replace(part, dest);
  return dest;
}
